### ISSUE TITLE
Insecure defaults

### ISSUE THREAD

**Issue by jasonroelofs** [Nedomas/databound#2]
`By default the permitted columns are all the columns in the database table.`

This is a very bad default. The default should follow Rails examples and be **no** columns at all, requiring the user to specify columns. If the defaults of this library insecure, the library will be used in an insecure way.

Also, I don't see any mention of permissions checking. What's to stop a user from running `User.delete(random_id)` or otherwise searching, retrieving, and updating resources that don't belong to the current user?

**Comment by Nedomas** [comment]
I'll have to agree. We just need to find a way to keep it simple.

**Permitted columns**

We can make it have no default permitted columns (https://github.com/Nedomas/databound-rails/blob/master/lib/databound/manager.rb#L58), but I'm not sure how to do it in a elegant way so we could just specify the route and have controller on the fly.
- One way is to have a custom controller generator from console. 
- The other is to have it specified in the `routes.rb`. But I'm not sure its the right place to be specifying models there.

**About the `delete`**

On the `create` or `update` part you can specify the DSL, like:

``` javascript
Post.create({ user_id: 'current', title: 'Hey' }).then(...
```

And handle it on the backend securely (it will error if its not `current`):

``` ruby
dsl(:user_id, :current) { current_user.id }
```

But there's no way currently to forbid deleting any record. Maybe a `allow_destroy_if` method? Something like:

``` ruby
allow_destroy_if do |record|
  record.user_id == current_user.id  
end
```

Got any ideas on these or other approaches?

Thanks for raising it! :heart: :heart: :heart:

**Comment by jasonroelofs** [comment]
**Permitted Columns**

I would recommend hooking into Rails' own Strong Parameters rather than building your own. http://edgeguides.rubyonrails.org/action_controller_overview.html#strong-parameters

**Delete and Update**

The normal way to protect against these kind of attacks is to query and filter according to what the user is allowed to see. For example, instead of `Project.find(10).destroy` you call `current_user.projects.find(10).destroy`. Then if the `current_user` is not linked to Project 10, the query blows up with a not-found error, preventing any destroy. If anything you wouldn't be dealing with an `allow_destroy` type call but a `default_query_scope` configuration, instead of, say, the `#model` method you define on the controller.

``` ruby
class ProjectsController < ApplicationController
  include Databound

  def query_scope
    current_user.projects
  end
end
```

Hope that helps!

**Comment by Nedomas** [comment]
About the **delete and update**.

How could we deal with polymorphic scope?

Where a `project` might belong to the `user` or to the `client` via `context_id` (and hence the `user` to the `client`? I was thinking about moving the scope that is currently implemented in javascript to the backend.

Now it is we secured via strict dsl, but we could totally define an extra scope in the backend if we found an elegant solution.

**Comment by jasonroelofs** [comment]
Well that's where this type of solution gets really tricky, because you really can't build a solution for every type of application or requirement. The least intrusive way I can think of is to make sure that if the user redefines the normal CRUD actions in their controller that those are called and not the handlers in Databound. Then users can let the default work when it can, and build custom as they need.

**Comment by Nedomas** [comment]
True. I believe there is a clean way not become the second ActiveScaffold.

I've implemented the `permit_update_destroy? do |record|` method  which allows us to handle both polymorphic and simple relations for now. So the security part is fixed now and is production-ready.

It is released in RubyGems with `databound 1.1.0`

https://github.com/Nedomas/databound-rails/commit/04e28afaf0d37f32241800425c98e809e1fbb9e9

What we need to do now is to find the way to implement secure defaults in such way that accomplishes both:
- by default you should specify the columns
- you could easily permit all columns without creating a controller.

An idea I've got right now:
- no columns are permitted by default
- but you can specify in `config/routes.rb`:

``` ruby
databound :users, permitted_columns: :table_columns
# or
databound :users, permitted_columns: %i(name)
```

I think this is borderline thing we can do in the routes.

**Comment by Nedomas** [comment]
Fixed, implemented and closed by https://github.com/Nedomas/databound-rails/commit/f1819a77b1c2f1d3421c8c89b6f4431bc82cbb7e

Thank you, Jason :shipit:
